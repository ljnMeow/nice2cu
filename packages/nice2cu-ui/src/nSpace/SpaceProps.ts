import { ExtractPropTypes, PropType } from 'vue';

function directionValidator(direction: string): boolean {
	return ['row', 'column'].includes(direction);
}
function justifyValidator(justify: string): boolean {
	return ['flex-start', 'flex-end', 'center', 'space-between', 'space-around'].includes(justify);
}
function alignValidator(align: string): boolean {
	return ['flex-start', 'center', 'flex-end', 'baseline'].includes(align);
}

export const SpaceProps = {
	direction: {
		type: String as PropType<'row' | 'column'>,
		default: 'row',
		validator: directionValidator,
	},
	justify: {
		type: String as PropType<'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around'>,
		default: 'flex-start',
		validator: justifyValidator,
	},
	align: {
		type: String as PropType<'flex-start' | 'center' | 'flex-end' | 'baseline'>,
		default: 'flex-start',
		validator: alignValidator,
	},
	wrap: {
		type: Boolean,
		default: true,
	},
	block: {
		type: Boolean,
		default: true,
	},
	size: {
		type: Array<string | number>,
		default: ['10px', '10px'],
	},
};

export type SpacePropsType = ExtractPropTypes<typeof SpaceProps>;
