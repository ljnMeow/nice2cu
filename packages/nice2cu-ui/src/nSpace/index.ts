import _space from './Space';
import { withInstall } from '../../utils/withInstall';

const nSpace = withInstall(_space);

export default nSpace;

export { nSpace };

declare module 'vue' {
	export interface GlobalComponents {
		nSpace: typeof nSpace;
	}
}
