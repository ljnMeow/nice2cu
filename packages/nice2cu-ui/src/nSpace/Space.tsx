import { computed, defineComponent, CSSProperties, VNode, Fragment } from 'vue';
import { createNamespace } from '../../utils/create';
import { handleUnit, isArray } from '../../utils/tools';
import { SpaceProps, SpacePropsType } from './SpaceProps';
import './style/space.less';

export default defineComponent({
	name: 'NSpace',
	props: SpaceProps,
	setup(props: SpacePropsType, { slots }) {
		const bem = createNamespace('space');

		const spaceStyle = computed<CSSProperties>(() => {
			return {
				flexDirection: props.direction,
				justifyContent: props.justify,
				alignItems: props.align,
				flexWrap: props.wrap ? 'wrap' : 'nowrap',
			};
		});

		const marginStyle = (index: number, lastIndex: number) => {
			const style: CSSProperties = {};

			const [horizontal, vertical] = props.size.map((item: string | number) => handleUnit(item));

			if (props.direction === 'row') {
				if (['flex-start', 'center', 'flex-end'].includes(props.justify)) {
					if (index !== lastIndex) {
						style.margin = `calc(${vertical} / 2) ${horizontal} calc(${vertical} / 2) 0`;
					} else {
						style.margin = `calc(${vertical} / 2) 0`;
					}
				} else if (props.justify === 'space-around') {
					style.margin = `calc(${vertical} / 2) calc(${horizontal} / 2)`;
				} else if (props.justify === 'space-between') {
					if (index === 0) {
						style.margin = `calc(${vertical} / 2) calc(${horizontal} / 2) calc(${vertical} / 2) 0`;
					} else if (index === lastIndex) {
						style.margin = `calc(${vertical} / 2) 0 calc(${vertical} / 2) calc(${horizontal} / 2)`;
					} else {
						style.margin = `calc(${vertical} / 2) calc(${horizontal} / 2)`;
					}
				}
			}

			if (props.direction === 'column' && index !== lastIndex) {
				style.margin = `0 0 ${vertical} 0`;
			}

			return style;
		};

		const filterFragment = (children: VNode[] = []) => {
			const result: VNode[] = [];

			children.forEach((vNode: any) => {
				if (vNode.type === Comment) {
					return;
				}

				if (vNode.type === Fragment && isArray(vNode.children)) {
					vNode.children.forEach((item: VNode) => {
						result.push(item);
					});
					return;
				}

				result.push(vNode);
			});

			return result;
		};

		return () => {
			const childrenList = filterFragment(slots && slots.default ? slots.default() : []);

			return (
				<div class={[bem.b(), props.block ? bem.m('block') : '', !props.wrap ? bem.m('scroll') : '']} style={spaceStyle.value}>
					{childrenList.map((child: VNode, index: number) => {
						const style: CSSProperties = marginStyle(index, childrenList.length - 1);

						return (
							<div class={bem.e('margin-box')} data-key={index} style={style}>
								{child}
							</div>
						);
					})}
				</div>
			);
		};
	},
});
