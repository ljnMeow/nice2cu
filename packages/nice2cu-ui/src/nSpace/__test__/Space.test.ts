import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import { createApp, h } from 'vue';

import nSpace from '../index';
import nSpaceVue from '../Space';

test('test Space plugin', () => {
	const app = createApp({}).use(nSpace);
	expect(app.component((nSpace as { name: string }).name)).toBeTruthy();
});

describe('test Space props', () => {
	test('test Space direction', () => {
		const directionArr: string[] = ['row', 'column'];
		directionArr.forEach((direction: string) => {
			const wrapper = mount(nSpaceVue as any, {
				props: {
					direction,
				},
			});
			expect(wrapper.find('.n-space').attributes('style')).toContain(`flex-direction: ${direction}`);
			wrapper.unmount();
		});
	});
	test('test Space justify', () => {
		const justifyArr: string[] = ['flex-start', 'flex-end', 'center', 'space-between', 'space-around'];
		justifyArr.forEach((justify: string) => {
			const wrapper = mount(nSpaceVue as any, {
				props: {
					justify,
				},
			});
			expect(wrapper.find('.n-space').attributes('style')).toContain(`justify-content: ${justify}`);
			wrapper.unmount();
		});
	});
	test('test Space justify', () => {
		const alignArr: string[] = ['flex-start', 'center', 'flex-end', 'baseline'];
		alignArr.forEach((align: string) => {
			const wrapper = mount(nSpaceVue as any, {
				props: {
					align,
				},
			});
			expect(wrapper.find('.n-space').attributes('style')).toContain(`align-items: ${align}`);
			wrapper.unmount();
		});
	});
	test('test Space wrap', async () => {
		const wrapper = mount(nSpaceVue, {
			props: {
				wrap: false,
			},
		});
		expect(wrapper.find('.n-space').attributes('style')).toContain(`flex-wrap: nowrap`);

		await wrapper.setProps({ wrap: true });
		expect(wrapper.find('.n-space').attributes('style')).toContain(`flex-wrap: wrap`);
		wrapper.unmount();
	});
	test('test Space block', async () => {
		const wrapper = mount(nSpaceVue, {
			props: {
				block: true,
			},
		});
		expect(wrapper.find('.n-space--block').exists()).toBe(true);

		await wrapper.setProps({ block: false });
		expect(wrapper.find('.n-space--block').exists()).toBe(false);
		wrapper.unmount();
	});
	test('test space default slots', () => {
		const wrapper = mount(nSpaceVue, {
			slots: {
				default: 'This is default slots',
			},
		});
		expect(wrapper.find('.n-space__margin-box').html()).toContain('This is default slots');
		wrapper.unmount();
	});
	test('test Space size', async () => {
		const wrapper = mount(nSpaceVue, {
			props: {
				size: [30, 30],
			},
			slots: {
				default: () => [1, 2, 3, 4].map((span) => h('div', [span])),
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
		wrapper.unmount();
	});
});
