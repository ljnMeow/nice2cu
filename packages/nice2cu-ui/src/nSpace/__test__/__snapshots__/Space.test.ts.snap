// Vitest Snapshot v1

exports[`test Space props > test Space size 1`] = `
"<div class=\\"n-space n-space--block\\" style=\\"flex-direction: row; justify-content: flex-start; align-items: flex-start; flex-wrap: wrap;\\">
  <div class=\\"n-space__margin-box\\" data-key=\\"0\\" style=\\"\\">
    <div>1</div>
  </div>
  <div class=\\"n-space__margin-box\\" data-key=\\"1\\" style=\\"\\">
    <div>2</div>
  </div>
  <div class=\\"n-space__margin-box\\" data-key=\\"2\\" style=\\"\\">
    <div>3</div>
  </div>
  <div class=\\"n-space__margin-box\\" data-key=\\"3\\" style=\\"\\">
    <div>4</div>
  </div>
</div>"
`;
